File: tests/goldens/typescript/valid_billing_policy.test.ts
import { describe, expect, test } from "vitest";
import { quoteSubscription } from "billing-policy";

describe("Billing policy contract", () => {
  test("grandfathers loyal customers onto the pro cap", () => {
    const result = quoteSubscription({
      account: {
        plan: "legacy",
        yearsActive: 7
      },
      usage: {
        projects: 18,
        seats: 4
      }
    });
    expect(result).toEqual({
      tier: "pro",
      monthlyUsd: 49,
      requiresReview: false
    });
  });
  test("flags enterprise usage before charging", () => {
    const result = quoteSubscription({
      account: {
        plan: "team",
        yearsActive: 1
      },
      usage: {
        projects: 91,
        seats: 42
      }
    });
    expect(result).toEqual({
      tier: "enterprise",
      monthlyUsd: null,
      requiresReview: true,
      reason: "seat_count"
    });
  });
});

File: tests/goldens/go/valid_billing_policy_test.go
package billingpolicy

import (
    "reflect"
    "testing"
)

// Source: tests/fixtures/valid_billing_policy.tdd:11
func TestGrandfathersLoyalCustomersOntoTheProCap(t *testing.T) {
    input := map[string]interface{}{"account": map[string]interface{}{"plan": "legacy", "yearsActive": 7}, "usage": map[string]interface{}{"projects": 18, "seats": 4}}
    result := quoteSubscription(input)

    expected := map[string]interface{}{"tier": "pro", "monthlyUsd": 49, "requiresReview": false}
    if !reflect.DeepEqual(result, expected) {
        t.Errorf("expected %v, got %v", expected, result)
    }
}

// Source: tests/fixtures/valid_billing_policy.tdd:25
func TestFlagsEnterpriseUsageBeforeCharging(t *testing.T) {
    input := map[string]interface{}{"account": map[string]interface{}{"plan": "team", "yearsActive": 1}, "usage": map[string]interface{}{"projects": 91, "seats": 42}}
    result := quoteSubscription(input)

    expected := map[string]interface{}{"tier": "enterprise", "monthlyUsd": nil, "requiresReview": true, "reason": "seat_count"}
    if !reflect.DeepEqual(result, expected) {
        t.Errorf("expected %v, got %v", expected, result)
    }
}
